src/storage-client/src/storage_collections.rs

```diff
@@ -1044,6 +1161,87 @@ where
+        // if we ever forget that we're not meant to use this code path.
+        tracing::error!("coming out of read-only mode gracefully, in-process, is not meant to be used in production");
+
+        // This is a carefully choreographed dance!
+
+        // First acquire a lock on collections, so no-one can create any while
+        // we operate.
+        let collections_to_upgrade = {
+            let self_collections = self.collections.lock().expect("lock poisoned");
+
+            // Check if we're already switched over.
+            if !self.read_only.load(Ordering::SeqCst) {
+                // Already done!
+                return;
+            }
+
+            // Pull a copy of all collections whose leased handle we have to
+            // "upgrade" to a critical handle.
+            let mut collections_to_upgrade = Vec::new();
+
+            for (id, collection) in self_collections.iter() {
+                collections_to_upgrade.push((
+                    *id,
+                    collection.collection_metadata.data_shard.clone(),
+                    collection.read_capabilities.frontier().to_owned(),
+                ));
+            }
+
+            // Record that we're out of read-only mode.
+            self.read_only.store(false, Ordering::SeqCst);
+
+            // And only then allow adding collections again! This will make sure
+            // that any collections registered after this time will already have
+            // critical since handles.
+            drop(self_collections);
+
+            collections_to_upgrade
+        };
+
+        let persist_client = self
+            .persist
+            .open(self.persist_location.clone())
+            .await
+            .unwrap();
+
+        for (id, data_shard, since) in collections_to_upgrade {
+            let since_handle = self
+                .open_critical_handle(&id, data_shard, Some(&since), &persist_client)
+                .await;
+
+            // NOTE(aljoscha): This can happen when the "active" `environmentd`
+            // keeps driving forward the critical handle. If we wanted this to
+            // be guaranteed to work, we'd need a mechanism from persist that
+            // allows turning back a critical handle's since to a since that is
+            // held by a leased reader. In theory, this is possible, because the
+            // overall shard since is being held back by the leased handle that
+            // we hold here.
+            //
+            // We *shouldn't* invest into that for now, because the approach we
+            // use for zero-downtime upgrades does not require transitioning out
+            // of read-only mode within the process. We instead halt the new
+            // `environmentd` and let it come up fresh in read-write mode, at
+            // which time it will take over the critical since handles right at
+            // the beginning, as per the usual bootstrap process.
```

Comment:
I didn't follow why halting the new envd and taking over the since handles that way will not be problematic. In the code below we panic if this environment cannot replace its read handle with a since handle at the same since. A command using the since of the leased read handle might be in transit to a cluster and so halting the new environment and restarting it doesn't help with the fact that we lost our since.

Reply:
In the expected case things will work out fine:

- when not yet halt!-ing, we hold our leased reader, this allows dataflows to be installed
- when halt!-ing with a compute command in-flight, the cluster has 15 minutes (the default lease timeout) to install it's dataflows, after that it might panic

But yes, there is potential for crashing when the cluster is stuck for more than 15 minutes and the handle expires. In the first version, we're not aiming for 100% solving everything, and in the expected case we _will_ improve downtime by quite a bit.